### PR TITLE
[Ready] Series of doc updates 

### LINKED ISSUE #239: confusing error message on trying to create rule with shared action
Trying this in the CLI

```
$ wsk rule create ifGitThenEcho myGitTrigger /whisk.system/samples/echo
error: The request content was malformed:
entity name malformed (code 1385580)
```

In `wsk rule create name trigger action` trigger or action can't be private
User needs to copy the action to his space like

```
$ wsk action create echo --copy /whisk.system/samples/echo
ok: created action echo
```

CLI error should be more clear about this and the [docs](https://github.com/openwhisk/openwhisk/blob/master/docs/triggers_rules.md#using-rules-to-associate-triggers-and-actions) should be updated to reflect this

### LINKED ISSUE #240: Update github package webhook trigger with sample output
Update the docs in [catalog](https://github.com/openwhisk/openwhisk/blob/master/docs/catalog.md#using-the-github-package)

To add a sample JSON object what the trigger emits like
```json
{
    "repository": {
        "tags_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/tags",
        "statuses_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/statuses/{sha}",
        "has_downloads": true,
        "blobs_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/git/blobs{/sha}",
        "git_refs_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/git/refs{/sha}",
        "issue_events_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/issues/events{/number}",
        "name": "openwhisk-github-integration",
        "has_issues": true,
        "watchers_count": 0,
        "forks": 0,
        "private": false,
        "size": 5,
        "open_issues_count": 0,
        "open_issues": 0,
        "subscribers_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/subscribers",
        "stargazers_count": 0,
        "url": "https://github.com/csantanapr/openwhisk-github-integration",
        "full_name": "csantanapr/openwhisk-github-integration",
        "releases_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/releases{/id}",
        "description": "Repository to demo openwhisk triggers and actions",
        "trees_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/git/trees{/sha}",
        "branches_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/branches{/branch}",
        "pushed_at": 1461186768,
        "git_url": "git://github.com/csantanapr/openwhisk-github-integration.git",
        "collaborators_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/collaborators{/collaborator}",
        "subscription_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/subscription",
        "languages_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/languages",
        "has_wiki": true,
        "commits_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/commits{/sha}",
        "contents_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/contents/{+path}",
        "fork": false,
        "git_tags_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/git/tags{/sha}",
        "downloads_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/downloads",
        "svn_url": "https://github.com/csantanapr/openwhisk-github-integration",
        "milestones_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/milestones{/number}",
        "id": 56718386,
        "compare_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/compare/{base}...{head}",
        "notifications_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/notifications{?since,all,participating}",
        "comments_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/comments{/number}",
        "pulls_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/pulls{/number}",
        "teams_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/teams",
        "forks_count": 0,
        "merges_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/merges",
        "stargazers": 0,
        "keys_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/keys{/key_id}",
        "deployments_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/deployments",
        "master_branch": "master",
        "contributors_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/contributors",
        "forks_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/forks",
        "clone_url": "https://github.com/csantanapr/openwhisk-github-integration.git",
        "created_at": 1461183654,
        "hooks_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/hooks",
        "ssh_url": "git@github.com:csantanapr/openwhisk-github-integration.git",
        "owner": {
            "email": "csantana23@gmail.com",
            "name": "csantanapr"
        },
        "html_url": "https://github.com/csantanapr/openwhisk-github-integration",
        "archive_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/{archive_format}{/ref}",
        "default_branch": "master",
        "updated_at": "2016-04-20T20:20:54Z",
        "issues_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/issues{/number}",
        "assignees_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/assignees{/user}",
        "events_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/events",
        "watchers": 0,
        "issue_comment_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/issues/comments{/number}",
        "labels_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/labels{/name}",
        "has_pages": false,
        "git_commits_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/git/commits{/sha}",
        "stargazers_url": "https://api.github.com/repos/csantanapr/openwhisk-github-integration/stargazers"
    },
    "before": "cd9acb00210e7badf191bf7243d6572259d972f5",
    "pusher": {
        "email": "csantana23@gmail.com",
        "name": "csantanapr"
    },
    "ref": "refs/heads/master",
    "commits": [
        {
            "author": {
                "username": "csantanapr",
                "email": "csantana@us.ibm.com",
                "name": "Carlos Santana"
            },
            "timestamp": "2016-04-20T17:12:42-04:00",
            "url": "https://github.com/csantanapr/openwhisk-github-integration/commit/b8a6922f88d1a83999a02f083de0674ba9183141",
            "tree_id": "01f437c196b3c7561c49d30a7856cecdf399d2e9",
            "removed": [],
            "modified": [
                "README.md"
            ],
            "added": [],
            "id": "b8a6922f88d1a83999a02f083de0674ba9183141",
            "distinct": true,
            "message": "update readme to trigger event",
            "committer": {
                "username": "csantanapr",
                "email": "csantana@us.ibm.com",
                "name": "Carlos Santana"
            }
        }
    ],
    "after": "b8a6922f88d1a83999a02f083de0674ba9183141",
    "forced": false,
    "sender": {
        "gists_url": "https://api.github.com/users/csantanapr/gists{/gist_id}",
        "organizations_url": "https://api.github.com/users/csantanapr/orgs",
        "gravatar_id": "",
        "url": "https://api.github.com/users/csantanapr",
        "repos_url": "https://api.github.com/users/csantanapr/repos",
        "received_events_url": "https://api.github.com/users/csantanapr/received_events",
        "id": 1094878,
        "following_url": "https://api.github.com/users/csantanapr/following{/other_user}",
        "site_admin": false,
        "subscriptions_url": "https://api.github.com/users/csantanapr/subscriptions",
        "starred_url": "https://api.github.com/users/csantanapr/starred{/owner}{/repo}",
        "html_url": "https://github.com/csantanapr",
        "login": "csantanapr",
        "type": "User",
        "events_url": "https://api.github.com/users/csantanapr/events{/privacy}",
        "avatar_url": "https://avatars.githubusercontent.com/u/1094878?v=3",
        "followers_url": "https://api.github.com/users/csantanapr/followers"
    },
    "head_commit": {
        "author": {
            "username": "csantanapr",
            "email": "csantana@us.ibm.com",
            "name": "Carlos Santana"
        },
        "timestamp": "2016-04-20T17:12:42-04:00",
        "url": "https://github.com/csantanapr/openwhisk-github-integration/commit/b8a6922f88d1a83999a02f083de0674ba9183141",
        "tree_id": "01f437c196b3c7561c49d30a7856cecdf399d2e9",
        "removed": [],
        "modified": [
            "README.md"
        ],
        "added": [],
        "id": "b8a6922f88d1a83999a02f083de0674ba9183141",
        "distinct": true,
        "message": "update readme to trigger event",
        "committer": {
            "username": "csantanapr",
            "email": "csantana@us.ibm.com",
            "name": "Carlos Santana"
        }
    },
    "compare": "https://github.com/csantanapr/openwhisk-github-integration/compare/cd9acb00210e...b8a6922f88d1",
    "deleted": false,
    "created": false
}
```

### LINKED ISSUE #306: Remove dash "-" from table in reference.md
Since "_" underscore can be use to denote the default namespace. 
The following table in https://github.com/openwhisk/openwhisk/blob/master/docs/reference.md#fully-qualified-names

| Fully qualified name | Alias | Namespace | Package | Name |
| --- | --- | --- | --- | --- |
| `/whisk.system/cloudant/read` | - | `/whisk.system` | `cloudant` | `read` |
| `/myOrg/video/transcode` | `video/transcode` | `/myOrg` | `video` | `transcode` |
| `/myOrg/filter` | `filter` | `/myOrg` | - | `filter` |

using dash `-` in the table can create confusion to the reader .
dash `-` might look like underscore `_` which it's meaningful for namespace location
  
It should be "Not Applicable", "None", or "Not Define"

### LINKED ISSUE #307: document the use of `_` for default namespace
Update docs to denote that the use of `_` underscore for namespace value is the same as using the default namespace which in bluemix is email (i.e. user@example.com) outside of bluemix will be the subject field
https://github.com/openwhisk/openwhisk/blob/master/docs/reference.md#namespaces-and-packages

### LINKED ISSUE #340: Update reference doc with action source code and parameters size limit
[Reference](https://github.com/openwhisk/openwhisk/blob/master/docs/reference.md#system-limits) is missing the following limits:

When creating or updating an action payload can't be larger than 1MB
Any activations (i.e. invoke action or fire a trigger) the payload can't be larger than 1MB

For is suggested that they use a tool concatenate all source code including dependencies into a single file when creating an action like suggested in this [post](https://developer.ibm.com/openwhisk/2016/03/17/bundling-openwhisk-actions-with-webpack/) for nodejs

Take into account that this is temporary as future implementations will remove some limits #254

### LINKED ISSUE #342: Document files and process limits for docker actions
Update [System Limits](https://github.com/openwhisk/openwhisk/blob/master/docs/reference.md#system-limits) with docker limits
```
val fileHandleLimit = Array("--ulimit", "nofile=64:64")
val processLimit = Array("--ulimit", "nproc=512:512")
```
Above info comes from : https://github.com/openwhisk/openwhisk/blob/master/core/dispatcher/src/whisk/core/container/ContainerUtils.scala#L71

Any other to document @rabbah @perryibm ?

### LINKED ISSUE #682: Using wrong type of quotes on code snippets in docs
Some code snippets use “ instead of "

### LINKED ISSUE #683: update actions docs to consistent use params and not msg

### LINKED ISSUE #684: docs missing step to install pod for ios starter app

### LINKED ISSUE #710: The link to API reference in SWagger doesn't work
There is a link to API reference in `docs/reference.md`.
I find the url doesn't work now in Swagger.
[API reference](http://petstore.swagger.io/?url=https://raw.githubusercontent.com/openwhisk/openwhisk/master/core/controller/src/resources/whiskswagger.json)

@csantanapr please take a look.

### PR DESCRIPTION
**Pull request by csantanapr**
- Update docs for ulmit limits for docker actions
  - Fixes #342
- Update docs with limits for action artifact and parameters
  - Fixes #340
- Update docs to define BASE URL and default namespace _
  - Fixes #307 #551
- Remove dash - since it can get confuse with default namespace underscore _
  - Fixes #306 
- Update docs on github payload content
  - Fixes #240
- Document restriction on creating triggers and rules
  - Fixes #239
- Update docs wrong type of quotes in code snippets
  - Fixes #682
- Update actions docs to consistent use params and not msg
  - Fixes #683
- Docs missing step to install pod for ios starter app 
  - Fixes #684
- Update version of nodejs from 0.12.9 to the correct version 0.12.14
- Update API url for swagger.json
  - Fixes #710

### CODE REVIEW

**Comment by csantanapr** [comment]
@rabbah please review when you have a chance

File: docs/catalog.md [hunk 1]
```diff
@@ -529,3 +528,7 @@ The following is an example of creating a trigger that will be fired each time t
   ```
   $ wsk trigger create myGitTrigger --feed myGit/webhook --param events push
   ```
+
+When you `push` a commit into the Github repository the trigger will be fired by the webhook, any action associated with a rule with the trigger will be invoked
```
Comment (rabbah) [inline docs/catalog.md]: A commit to the Github repository via a `git push` will cause the trigger to be fired by the webhook. If there is a rule that matches the trigger, then the associated action will be invoke. The action receives the Github webhook payload as an input parameter. Each Github webhook event has a similar JSON schema, but a unique `payload` object that is determined by its event type.

File: docs/reference.md [hunk 2]
```diff
@@ -359,13 +364,30 @@ OpenWhisk has a few system limits, including how much memory an action uses and
 * A user can change the limit when creating the action.
 * A container cannot have more memory allocated than the limit.
 
-### Per namespace #Concurrent Invocation (#) (Default: 100)
+### Per action artifact (MB) (Fixed: 1MB)
+* The maximum size for the artifact (e.g., file name) containing action definition is 1 MB.
```
Comment (rabbah) [inline docs/reference.md]: file name -> file (or code size)

File: docs/reference.md [hunk 3]
```diff
@@ -359,13 +364,30 @@ OpenWhisk has a few system limits, including how much memory an action uses and
...
+* It's recommended for nodejs action to use a tool to concatenate all source code including dependencies into a single bundled file
```
Comment (rabbah) [inline docs/reference.md]: Node.js

File: docs/reference.md [hunk 4]
```diff
@@ -359,13 +364,30 @@ OpenWhisk has a few system limits, including how much memory an action uses and
 * A user can change the limit when creating the action.
 * A container cannot have more memory allocated than the limit.
 
-### Per namespace #Concurrent Invocation (#) (Default: 100)
+### Per action artifact (MB) (Fixed: 1MB)
+* The maximum size for the artifact (e.g., file name) containing action definition is 1 MB.
+* It's recommended for nodejs action to use a tool to concatenate all source code including dependencies into a single bundled file 
+
+### Per activation payload size (MB) (Fixed: 1MB)
+* Any activations (i.e. invoke action or fire a trigger) the payload can't be larger than 1MB
```
Comment (rabbah) [inline docs/reference.md]: The maximum `POST` content size for an action invokation or trigger firing is 1MB.

File: docs/reference.md [hunk 5]
```diff
@@ -359,13 +364,30 @@ OpenWhisk has a few system limits, including how much memory an action uses and
...
+* The payload includes parameters for the action or trigger including strings or encoded binary data
```
Comment (rabbah) [inline docs/reference.md]: Strike as not necessary?

File: docs/reference.md [hunk 6]
```diff
@@ -359,13 +364,30 @@ OpenWhisk has a few system limits, including how much memory an action uses and
...
-### Invocations per minute/hour (#) (Fixed: 120/3600)
+### Invocations per minute/hour (Fixed: 120/3600)
 * The rate limit N is set to 120/3600 and limits the number of action invocations in one minute/hour windows.
 * A user cannot change this limit when creating the action.
 * A CLI call that exceeds this limit receives an error code corresponding to TOO_MANY_REQUESTS.
+
+### Per Docker action open files ulimit (Fixed: 64:64)
+* The maximum number of files open is 64 for hard limit and 64 for soft limit
+* The docker run command use the argument `--ulimit nofile=64:64`
+* For more information on the ulimit for open files see the [docker run](https://docs.docker.com/engine/reference/commandline/run) documentation
```
Comment (rabbah) [inline docs/reference.md]: add periods.

File: docs/reference.md [hunk 7]
```diff
@@ -359,13 +364,30 @@ OpenWhisk has a few system limits, including how much memory an action uses and
...
 
-
-### Invocations per minute/hour (#) (Fixed: 120/3600)
+### Invocations per minute/hour (Fixed: 120/3600)
 * The rate limit N is set to 120/3600 and limits the number of action invocations in one minute/hour windows.
 * A user cannot change this limit when creating the action.
 * A CLI call that exceeds this limit receives an error code corresponding to TOO_MANY_REQUESTS.
+
+### Per Docker action open files ulimit (Fixed: 64:64)
+* The maximum number of files open is 64 for hard limit and 64 for soft limit
```
Comment (rabbah) [inline docs/reference.md]: The maximum number of open file is 64 (this applies to both hard and soft limits).

File: docs/reference.md [hunk 8]
```diff
@@ -359,13 +364,30 @@ OpenWhisk has a few system limits, including how much memory an action uses and
...
+* The docker run command use the argument `--ulimit nofile=64:64`
+* For more information on the ulimit for open files see the [docker run](https://docs.docker.com/engine/reference/commandline/run) documentation
+
+### Per Docker action number of processes ulimit (Fixed: 512:512)
+* The maximum number of processes available to a user is 512 for hard limit and 512 for soft limit
```
Comment (rabbah) [inline docs/reference.md]: The maximum number of processes available to a user is 512 (this applies to both hard and soft limits).

File: docs/reference.md [hunk 9]
```diff
@@ -359,13 +364,30 @@ OpenWhisk has a few system limits, including how much memory an action uses and
...
+* The docker run command use the argument `--ulimit nproc=512:512`
+* For more information on the ulimit for maximum number of processes see the [docker run](https://docs.docker.com/engine/reference/commandline/run) documentation
```
Comment (rabbah) [inline docs/reference.md]: periods for above two lines.

File: docs/triggers_rules.md [hunk 10]
```diff
@@ -82,6 +82,7 @@ As an example, create a trigger to send user location updates, and manually fire
 
    Any events that you fire to the statusUpdate trigger currently don't do anything. To be useful, the trigger needs a rule that associates it with an action.
 
+Triggers can't be created inside a package, they must be created directly under a namespace.
```
Comment (rabbah) [inline docs/triggers_rules.md]: Indent? can't be created inside a package -> may not be created inside a package. Triggers are only allowed in the default package. (Do we defined default package)?

File: docs/triggers_rules.md [hunk 11]
```diff
@@ -141,3 +142,6 @@ As an example, create a rule that calls the hello action whenever a location upd
   You see that the hello action received the event payload and returned the expected string.
 
   You can create multiple rules that associate the same trigger with different actions.
+
+The trigger and action that make a rule must be in the same namespace, which is in fact private.
```
Comment (rabbah) [inline docs/triggers_rules.md]: indent.

FYI this constraint should be lifted soon now that new activator is merged.

File: docs/triggers_rules.md (hunk 10 above)
Comment (csantanapr) [inline docs/triggers_rules.md]: I going to change to "Triggers may not be created inside a package, they must be created directly under a namespace."

>Triggers are only allowed in the default package. (Do we defined default package)?
What are you referring here by "default package"?

File: docs/triggers_rules.md (hunk 11 above)
Comment (csantanapr) [inline docs/triggers_rules.md]: I'm going to change to " The trigger and action that make a rule must be in the same namespace."

>and must belong to the default package
What does this means?
You want me to add it, I don't know what a default package is.

Comment (csantanapr) [inline docs/triggers_rules.md]: So we don't have this restriction any more?
If this is fixed I would prefer to remove it then.

So any restrictions now?
trigger can be created on any namespace, and the action can be created on any namespace and package, then the rule can be created with full qualified names to be sure the correct entities are specified?

**Comment by csantanapr** [comment]
@rabbah thanks for the review, I tried to address them and added a fix for the swagger url #710

File: docs/catalog.md [hunk 12]
```diff
@@ -533,3 +532,7 @@ The following is an example of creating a trigger that will be fired each time t
   ```
   $ wsk trigger create myGitTrigger --feed myGit/webhook --param events push
   ```
+
+A commit to the Github repository via a git push will cause the trigger to be fired by the webhook. If there is a rule that matches the trigger, then the associated action will be invoke.
```
Comment (rabbah) [inline docs/catalog.md]: be invoke -> be invoked

File: docs/reference.md [hunk 13]
```diff
@@ -359,13 +364,29 @@ OpenWhisk has a few system limits, including how much memory an action uses and
 * A user can change the limit when creating the action.
 * A container cannot have more memory allocated than the limit.
 
-### Per namespace #Concurrent Invocation (#) (Default: 100)
+### Per action artifact (MB) (Fixed: 1MB)
+* The maximum code size for the action is 1MB.
+* It's recommended for a JavaScript action to use a tool to concatenate all source code including dependencies into a single bundled file
```
Comment (rabbah) [inline docs/reference.md]: period.

File: docs/triggers_rules.md [hunk 14]
```diff
@@ -80,8 +80,9 @@ As an example, create a trigger to send user location updates, and manually fire
   ok: triggered locationUpdate with id fa495d1223a2408b999c3e0ca73b2677
   ```
 
-   Any events that you fire to the statusUpdate trigger currently don't do anything. To be useful, the trigger needs a rule that associates it with an action.
+  Any events that you fire to the statusUpdate trigger currently don't do anything. To be useful, the trigger needs a rule that associates it with an action.
```
Comment (rabbah) [inline docs/triggers_rules.md]: statusUpdate -> locationUpdate

**Comment by rabbah** [comment]
Made some small edits, rebased and merged as https://github.com/openwhisk/openwhisk/commit/2b6eb3da0570a73d9aa0c46bc5b96cebcd5d2ab5.

**Comment by csantanapr** [comment]
Thanks @rabbah

**Pull request closed without merging**